src/output.rs

```diff
@@ -1,2 +1,231 @@
+use std::{borrow::Borrow, fmt::Write};
+
 pub mod flat;
 pub mod tree;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Symbol {
+    Vacuum,
+    Elbow,
+    Pipe,
+    Junction,
+}
+
+impl std::fmt::Display for Symbol {
+    #[rustfmt::skip]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Vacuum   => f.write_str("    "),
+            Self::Pipe     => f.write_str("│   "),
+            Self::Junction => f.write_str("├── "),
+            Self::Elbow    => f.write_str("└── "),
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+pub enum Fragment {
+    #[default]
+    Empty,
+    Symbol(Symbol),
+    Char(char),
+    String(std::borrow::Cow<'static, str>),
+    DisplayAtom(std::rc::Rc<dyn std::fmt::Display>),
```

Comment:
I think can import Cow and Rc directly rather than qualifying them everywhere.

Reply:
Done.